source/common/network/cidr_range.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+#include "envoy/network/address.h"
+
+namespace Network {
+namespace Address {
+
+/*
+ * A "Classless Inter-Domain Routing" range of internet addresses, aka a CIDR range, consisting
+ * of an Ip address and a count of leading bits included in the mask. Other than those leading
+ * bits, all of the other bits of the Ip address are zero. For more info, see RFC1519 or
+ * https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing.
+ */
+class CidrRange {
+public:
+  /**
+   * Constructs an uninitialized range: length == -1, and there is no associated address.
+   */
+  CidrRange();
+
+  /**
+   * Copies an existing CidrRange.
+   */
+  CidrRange(const CidrRange& other);
+
+  /**
+   * Overwrites this with other.
+   */
+  CidrRange& operator=(const CidrRange& other);
+
+  /**
+   * @return Ipv4 address data IFF length >= 0 and version() == IpVersion::v4, otherwise nullptr.
+   */
+  const Ipv4* ipv4() const;
+
+  /**
+   * @return Ipv6 address data IFF length >= 0 and version() == IpVersion::v6, otherwise nullptr.
+   */
+  const Ipv6* ipv6() const;
+
+  /**
+   * @return the number of bits of the address that are included in the mask.
+   * -1 if uninitialized or invalid, else in the range 0 to 32 for IPv4, and 0 to 128 for IPv6.
+   */
+  int length() const;
+
+  /**
+   * @return the version of IP address.
+   */
+  IpVersion version() const;
+
+  /**
+   * @return true if the address argument is in the range of this object.
+   * false if not, including if the range is uninitialized or if the argument is not of the
+   * same IpVersion.
+   */
+  bool isInRange(const InstancePtr& address) const;
+
+  /**
+   * @return a human readable string for the range.
+   *
+   * This string will be in the following format:
```

Comment:
We don't currently run doxygen on the source (but we should). The comment format is loose and was intended to eventually work with doxygen / most editors that auto parse.